transform/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,233 @@
+
+func (k KubernetesTransformPlugin) getWhiteOuts(groupKind schema.GroupKind) bool {
+	if groupKind == endpointGK {
+		return true
+	}
+
+	if groupKind == endpointSliceGK {
+		return true
+	}
+
+	// For right now we assume PVC's are handled by a different part
+	// of the tool chain.
+	if groupKind == pvcGK {
+		return true
+	}
+	return false
+}
+
+func (k KubernetesTransformPlugin) getKubernetesTransforms(obj unstructured.Unstructured) (jsonpatch.Patch, error) {
+
+	// Always attempt to add annotations for each thing.
+	jsonPatch := jsonpatch.Patch{}
+	if len(k.AddedAnnotations) > 0 {
+		patches, err := addAnnotations(k.AddedAnnotations)
+		if err != nil {
+			return nil, err
+		}
+		jsonPatch = append(jsonPatch, patches...)
+	}
+	if podGK == obj.GetObjectKind().GroupVersionKind().GroupKind() {
+		patches, err := removePodSelectedNode()
+		if err != nil {
+			return nil, err
+		}
+		jsonPatch = append(jsonPatch, patches...)
+	}
+	if len(k.RegistryReplacement) > 0 {
+		if podGK == obj.GetObjectKind().GroupVersionKind().GroupKind() {
+			// jsonPatch for return
+		} else if template, ok := types.IsPodSpecable(obj); ok {
+			jps := jsonpatch.Patch{}
+			for i, container := range template.Spec.Containers {
+				updatedImage, update := updateImageRegistry(k.RegistryReplacement, container.Image)
+				if update {
+					jp, err := updateImage(fmt.Sprintf("/spec/template/spec/containers/%v/image", i), updatedImage)
+					if err != nil {
+						return nil, err
+					}
+					jps = append(jps, jp...)
+				}
+			}
+			for i, container := range template.Spec.InitContainers {
+				updatedImage, update := updateImageRegistry(k.RegistryReplacement, container.Image)
+				if update {
+					jp, err := updateImage(fmt.Sprintf("/spec/template/spec/initContainers/%v/image", i), updatedImage)
+					if err != nil {
+						return nil, err
+					}
+					jps = append(jps, jp...)
+				}
+			}
+			jsonPatch = append(jsonPatch, jps...)
+		}
+	}
+	if obj.GetObjectKind().GroupVersionKind().GroupKind() == serviceGK {
+		patches, err := removeServiceClusterIPs()
+		if err != nil {
+			return nil, err
+		}
+		jsonPatch = append(jsonPatch, patches...)
+	}
+
+	return jsonPatch, nil
+}
+
+func updateImageRegistry(registryReplacements map[string]string, oldImageName string) (string, bool) {
+	// Break up oldImage to get the registry URL. Assume all manifests are using fully qualified image paths, if not ignore.
+	imageParts := strings.Split(oldImageName, "/")
+	if len(imageParts) != 3 {
+		return "", false
+	}
+	if newRegistry, ok := registryReplacements[imageParts[0]]; ok {
+		return strings.Join([]string{newRegistry, imageParts[1], imageParts[2]}, "/"), true
+	}
+
+	return "", false
+}
+
+func addAnnotations(addedAnnotations map[string]string) (jsonpatch.Patch, error) {
+	patchJSON := `[`
+	i := 0
+	for key, value := range addedAnnotations {
+		if i == 0 {
+			patchJSON = fmt.Sprintf(`%v
+{"op": "add", "path": "/metadata/annotations/%v", "value": "%v"}`, patchJSON, key, value)
```

Comment:
nit: I wonder if we can collect these strings into one place and then read and reference those from there.

In the past I have seen collecting assets in a common place allows for more automation[1][2]. we are in such early stages not sure if this is applicable but just curious if you have any thoughts regarding this

1. https://github.com/openshift/cluster-etcd-operator/blob/master/bindata/etcd/pod-cm.yaml
2. automation https://github.com/openshift/cluster-etcd-operator/blob/2f6c869efbe534f3e44fc1732019488521386807/Makefile#L32

Reply:
I could see moving these to constants at the top of the file WDYT? 

Or do you see us needing to update this on particular builds based on something? I am not seeing that but could be missing it

Reply:
The top of the file seems fine for now.